src/modules/content-s3/s3.c

```diff
@@ -0,0 +1,268 @@
+ * Copyright 2020 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <sys/types.h>
+#include <string.h>
+#include <errno.h>
+#include <stdio.h>
+#include <libs3.h>
+#include <stdlib.h>
+
+#include "src/common/libutil/errno_safe.h"
+
+#include "s3.h"
+
+static S3Protocol protocol = S3ProtocolHTTP;
+static S3UriStyle uri_style = S3UriStylePath;
+
+/* Data needed by the get object callback function
+ */
+struct cb_data {
+    size_t size;
+    void *data;
+    int count;
+    S3Status status;
+};
+
+static S3Status response_props_cb (const S3ResponseProperties *properties, void *data)
+{
+    return S3StatusOK;
+}
+
+static void response_complete_cb (S3Status status, const S3ErrorDetails *error, void *data)
+{
+    struct cb_data *ctx = data;
+    ctx->status = status;
+}
+
+/* Writes a chunk of 'data' to s3, returning
+ * the size of the data written.
+ */
+static int put_object_cb (int buff_size, char *buff, void *data)
+{
+    struct cb_data *ctx = data;
+    int size = (buff_size < ctx->size - ctx->count ? buff_size : ctx->size - ctx->count);
+
+    memcpy (buff, ctx->data + ctx->count, size);
+    ctx->count += size;
+
+    return size;
+}
+
+/* Gets the object from s3 storing it in in 'data->datap' as well
+ * as updating 'data->sizep' to hold the size of the data read.
+ */
+static S3Status get_object_cb (int buff_size, const char *buff, void *data)
+{
+    void *tmp = NULL;
+    struct cb_data *ctx = data;
+
+    if (! (tmp = realloc (ctx->data, buff_size + ctx->size)))
+        return S3StatusOutOfMemory;
+    ctx->data = tmp;
+
+    memcpy (ctx->data + ctx->size, buff, buff_size);
+    ctx->size += buff_size;
+
+    return S3StatusOK;
+}
+
+int s3_init (struct s3_config *cfg, const char **errstr)
+{
+    S3Status status = S3_initialize ("s3", S3_INIT_ALL, cfg->hostname);
+
+    if (status != S3StatusOK) {
+        errno = ECONNREFUSED;
+        if (errstr)
+            *errstr = S3_get_status_name (status);
+
+        return -1;
+    }
+    
+    return 0;
+}
+
+void s3_cleanup ()
+{
+    S3_deinitialize ();
+}
+
+int s3_bucket_create (struct s3_config *cfg, const char **errstr)
+{
+    int retries = cfg->retries;
+    S3Status status = S3_validate_bucket_name (cfg->bucket, uri_style);
+
+    if (status != S3StatusOK) {
+        errno = ECONNREFUSED;
+        if (errstr)
+            *errstr = S3_get_status_name (status);
+
+        return -1;
+    }
+
+    S3ResponseHandler resp_hndl = {
+        .propertiesCallback = &response_props_cb,
+        .completeCallback = &response_complete_cb
+    };
+
+    struct cb_data ctx = {
+        0,
+        NULL,
+        0,
+        status,
```

Comment:
For consistency, use tagged initializers here too.  It also makes the code a bit easier to read.